controllers/restore.go

```diff
@@ -839,3 +845,164 @@ func setUserRestoreFilters(
 		}
 	}
 }
+
+// delete all velero restores created by the acmRestore object
+// and remove finalizers
+func finalizeRestore(
+	ctx context.Context,
+	c client.Client,
+	acmRestore *v1beta1.Restore,
+	veleroRestoreList veleroapi.RestoreList,
+) error {
+
+	reqLogger := log.FromContext(ctx)
+
+	for _, veleroRestore := range veleroRestoreList.Items {
+		if err := c.Delete(ctx, &veleroRestore); err != nil {
+			reqLogger.Error(err, fmt.Sprintf("Error terminating restore: %s", veleroRestore.GetName()))
+		}
+	}
+
+	if len(veleroRestoreList.Items) != 0 {
+		reqLogger.Info("Waiting for velero restores to be terminated")
+		return fmt.Errorf("waiting for velero restores to be terminated")
+	}
+
+	reqLogger.Info("Successfully finalized Restore " + acmRestore.GetName())
+	return nil
+}
+
+// return the internalhubcomponent with the cluster-backup name
+func getInternalHubResource(
+	ctx context.Context,
+	dyn dynamic.Interface,
+) (*unstructured.Unstructured, error) {
+
+	reqLogger := log.FromContext(ctx)
+	reqLogger.Info("get cluster-backup  internalhubcomponent")
+
+	mchGVRList := schema.GroupVersionResource{Group: ihcGroup,
+		Version: "v1", Resource: "internalhubcomponents"}
+
+	mchList, err := dyn.Resource(mchGVRList).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return &unstructured.Unstructured{}, err
+	}
+	//
+	reqLogger.Info("get cluster-backup  internalhubcomponent")
+	for i := range mchList.Items {
+		item := mchList.Items[i]
+
+		if item.GetName() == "cluster-backup" {
+			return &item, nil
+		}
+	}
+
+	return &unstructured.Unstructured{}, err
+
+}
+
+// remove the acm finalizer and
+// remove internalhubcomponent finalizer if this is the only restore resource found
+func removeResourcesFinalizer(
+	ctx context.Context,
+	c client.Client,
+	dyn dynamic.Interface,
+	internalHubResource *unstructured.Unstructured,
+	acmRestore *v1beta1.Restore,
+) error {
+
+	reqLogger := log.FromContext(ctx)
+
+	// Remove restore finalizer. Once all finalizers have been
+	// removed, the object will be deleted.
+	needsUpdate := controllerutil.RemoveFinalizer(acmRestore, acmRestoreFinalizer)
+
+	// if no other restore resources, remove mch finalizer
+	acmRestoreList := v1beta1.RestoreList{}
+	if err := c.List(
+		ctx,
+		&acmRestoreList,
+		client.InNamespace(acmRestore.GetNamespace())); err == nil &&
+		len(acmRestoreList.Items) == 1 {
+
+		// remove InternalHubResource restore finalizer if this is the last resource to be deleted
+		mchGVRList := schema.GroupVersionResource{
+			Group:   ihcGroup,
+			Version: "v1", Resource: "internalhubcomponents"}
+		if internalHubResource != nil {
+			if fins := internalHubResource.GetFinalizers(); fins != nil && findValue(fins, acmRestoreFinalizer) {
```

Comment:
Could we use controllerutil.RemoveFinalizer here? Rather than checking/removing?  Like you do above on line 919 for the acmRestore finalizer.

Reply:
this is an unstructured resource, can't use the controllerutil.RemoveFinalizer